bin/arrayNormalization.R

```diff
@@ -0,0 +1,374 @@
+		assayNames <- assaysToDataFiles$AssayName
+
+		# If data is from a 1-colour or 2-colour Agilent array, we do the
+		# pre-processing with limma. This consists of subsetting probes based
+		# on miRBase mappings, background correction, within- or between-array
+		# normalization, and averaging over duplicates.
+		# For 1-colour data, normalized intensities are written to a single file.
+		# For 2-colour data, log2(fols change)s are written to one file and
+		# average intensities to another file.
+		if(mode == "agil1" | mode == "agil2") {
+			
+			# load limma
+			library(limma)
+
+			# read files
+			if(mode == "agil1") {
+			
+				print("Reading Agilent 1-colour data")
+				# for 1-colour Agilent data we only have a green channel.
+				dataSet <- try({read.maimages(files, source="agilent", green.only=TRUE)})
+			}
+			else if(mode == "agil2") {
+				
+				print("Reading Agilent 2-colour data")
+				dataSet <- try({read.maimages(files, source="agilent")})
+			}
+			# check it worked, if not, return the error.
+			if(class(dataSet) == "try-error") { return(dataSet) }
+			
+			# Check if there's a microRNA miRBase probeset list and if so,
+			# subset based on that list.
+			if(miRBaseFile != 0) { dataSet <- subsetProbes(dataSet, miRBaseFile) }
+			
+			# Background correction.
+			# Using method outlined in Section 6.1 of Limma Users Guide (25
+			# September 2012 version), to adjust the foreground adaptively for
+			# the background intensities, to produce strictly positive adjusted
+			# intensities, i.e. avoiding negative or zero values in the
+			# resulting adjusted intensities.
+			# NB: this might get masked by oligo package's backgroundCorrect
+			# method if they are both loaded (here they shouldn't be so don't
+			# call limma's explicitly).
+			print("Background correcting")
+			dataSet <- backgroundCorrect(dataSet, method="normexp", offset=50)
+			
+			if(mode == "agil1") {
+				
+				# Between-array normalization for 1-colour.
+				normData <- normalizeBetweenArrays(dataSet, method="quantile")
+			}
+
+			else if(mode == "agil2") {
+				# Within-array normalization for 2-colour.
+				# This gives the M-value and A-value for each gene on the array.
+				# M = log2(Cy5) - log2(Cy3)
+				# A = 0.5[log2(Cy5) + log2(Cy3)]
+				print("Running within-array normalization")
+				normData <- normalizeWithinArrays(dataSet, method="loess")
+			}
+
+			# average for probe duplicates -- some probes are on the array
+			# multiple times in different locations.
+			print("Averaging duplicated probes")
+			normData <- avereps(normData, ID=normData$genes$ProbeName)
+			
+			# Write results to file
+			if(mode == "agil1") {
+				# For single channel Agilent data, this is just the
+				# quantile-normalized expressions (analogous to RMA-normalized
+				# expressions from Affymetrix data).
+				
+				# Make a dataframe with probe IDs and normalized expressions.
+				normExprs <- data.frame(cbind(normData$genes$ProbeName, normData$E))
```

Comment:
I think that you can avoid the cbind and the rename by simply doing:
```
normExprs <- data.frame(DesignElementAccession=normData$genes$ProbeName, assayNames=normData$E)
```

Reply:
I don't think so. Because `normData$E` is the expression values matrix. Here, we are renaming columns in the matrix with first column as `DesignElementAccession` and rest columns with assay names derived from `assayNames <- assaysToDataFiles$AssayName`

Reply:
ahh, I see what you mean, my bad. I thought it was a vector.